### ISSUE TITLE
FatalErrorException render loop

### ISSUE THREAD

**Issue by drkblog** [cakephp/cakephp#5595]
There is a loop whenever Cake 2.6 has to render a fatal error (like calling a non-existent method on a PHP object). The fatal error rendering triggers another fatal error until memory is exhausted. I rolled back to 2.5.8 and this issue disappeared.

Sample log:

    2015-01-07 09:55:38 Error: Fatal Error (256): [CakeException] Headers already sent in /home/leandro/htdocs/visitor/app/Controller/UsersController.php on line 37
    #0 /home/leandro/htdocs/visitor/lib/Cake/Network/CakeResponse.php(419): CakeResponse->_sendHeader('HTTP/1.1 500 In...')
    #1 /home/leandro/htdocs/visitor/lib/Cake/Error/ExceptionRenderer.php(327): CakeResponse->send()
    #2 /home/leandro/htdocs/visitor/lib/Cake/Error/ExceptionRenderer.php(306): ExceptionRenderer->_outputMessageSafe('error500')
    #3 /home/leandro/htdocs/visitor/lib/Cake/Error/ExceptionRenderer.php(210): ExceptionRenderer->_outputMessage('fatalError')
    #4 [internal function]: ExceptionRenderer->_cakeError(Object(FatalErrorException))
    #5 /home/leandro/htdocs/visitor/lib/Cake/Error/ExceptionRenderer.php(187): call_user_func_array(Array, Array)
    #6 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(119): ExceptionRenderer->render()
    #7 [internal function]: ErrorHandler::handleException(Object(FatalErrorException))
    #8 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(252): call_user_func('ErrorHandler::h...', Object(FatalErrorException))
    #9 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(203): ErrorHandler::handleFatalError(256, '[CakeException]...', '/home/leandro/h...', 128)
    #10 [internal function]: ErrorHandler::handleError(256, '[CakeException]...', '/home/leandro/h...', 128, Array)
    #11 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(128): trigger_error('[CakeException]...', 256)
    #12 [internal function]: ErrorHandler::handleException(Object(FatalErrorException))
    #13 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(252): call_user_func('ErrorHandler::h...', Object(FatalErrorException))
    #14 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(203): ErrorHandler::handleFatalError(256, '[CakeException]...', '/home/leandro/h...', 128)
    #15 [internal function]: ErrorHandler::handleError(256, '[CakeException]...', '/home/leandro/h...', 128, Array)
    #16 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(128): trigger_error('[CakeException]...', 256)
    ..............
    ..............
    #127 [internal function]: ErrorHandler::handleException(Object(FatalErrorException))
    #128 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(252): call_user_func('ErrorHandler::h...', Object(FatalErrorException))
    #129 /home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php(203): ErrorHandler::handleFatalError(1, 'Call to undefin...', '/home/leandro/h...', 37)
    #130 [internal function]: ErrorHandler::handleError(1, 'Call to undefin...', '/home/leandro/h...', 37, Array)
    #131 /home/leandro/htdocs/visitor/lib/Cake/Core/App.php(931): call_user_func('ErrorHandler::h...', 1, 'Call to undefin...', '/home/leandro/h...', 37, Array)
    #132 /home/leandro/htdocs/visitor/lib/Cake/Core/App.php(904): App::_checkFatalError()
    #133 [internal function]: App::shutdown()
    #134 {main} in [/home/leandro/htdocs/visitor/lib/Cake/Error/ErrorHandler.php, line 128]

**Comment by markstory** [comment]
I did a bisect on the differences between 2.5.8(good) and 2.6.0(bad). The problematic commit seems to be 792c251890d67ac643785def4ff5d8ee59318656. Reverting the exception throwing fixes the issue, but re-opens the issue described in #3990.

Given that we can't really set headers from the fatal error handlers as part of PHP's shutdown seems to send headers, I can look into ways of disabling the header errors.

**Comment by ADmad** [comment]
I would just revert the change done in #3990. It causes more problems than it solves. Another problem it causes is you can no longer stick a debug() in your controller for quick debugging as the exception kicks in and prevents view rendering.

**Comment by dereuromark** [comment]
What if we do it only for debug mode off and/or trigger a notice instead? I found the exception too harsh. But a trigger_error() would still be useful.

**Comment by ADmad** [comment]
Throwing exception when debug is on and not thing anything when off sounds good to me.

**Comment by markstory** [comment]
I will change the code to trigger error instead.

**Comment by pyrite357** [comment]
Is it possible you can disable this commit when autoRender is set to false, and not just debug off? This commit prevented me from upgrading to 2.6.0 and I had to rollback to 2.5.8 because I have ajax methods throughout my site that look like this:

```php
public function getsomedata() {
    Configure::write('debug', 1);
    $this->autoRender = false;
    $data = $this->SomeModel->getSomeData('someparam');
    echo $data;
}
```
and plenty of others like this:

```php
public function getsomedata2() {
    Configure::write('debug', 1);
    $this->autoRender = false;
    $data = $this->SomeModel->query("SELECT * FROM posts");
    echo json_encode($data);
}
```

Now I know with 2.6.0, if I use jsonViews for the 2nd method, problem solved, but I have hundreds of methods I would have to upgrade to jsonViews. The first method though, I don't know what is causing the headers to be already sent? And I don't know a way around it other than using ```exit()``` at the end.

I've been using CakePHP for my company since 2008, and my codebase has over 300,000 lines of code, and is difficult to maintain across upgrades with major changes like this that break my site. Sure as of 2.4 or whatever I should have started using jsonViews and what have you. But developers shouldn't be penalized for not using new features. Prior to that I had already implemented my own version of jsonViews, and the majority of my codebase is using it. But that causes this error on 2.6.0. Thank you for listening, sorry for the long post.

**Comment by dereuromark** [comment]
@pyrite357 Why does that need a `Configure::write('debug', 1);`? I don't see a reason for that.

**Comment by pyrite357** [comment]
I was just illustrating what is in my core config.

**Comment by dereuromark** [comment]
@pyrite357 I have different feelings about this. On the one hand, I understand your issue and that, even though you want to upgrade to the next minor release you do not want to modify much code.

On the other hand, echoing directly in your controller never really was something you should have ever done in 2.x (not even in 2.0). And as such, asking for not breaking invalid code seems like a lot somehow, whereas you can either not (yet) upgrade to 2.6 or easily refactor your controller action code:
```php
$this->autoRender = false;
$data = $this->SomeModel->query("SELECT * FROM posts");
$this->response->body(json_encode($data));
```
etc. (really minor change).

**Comment by pyrite357** [comment]
I think all I'm asking for is if we can do what @ADmad said with "Throwing exception when debug is on and not thing anything when off sounds good to me." and change it to "Throwing exception when debug is on and not thing anything when off **or autoRender is false** sounds good to me."

**Comment by pyrite357** [comment]
@dereuromark I understand, and your logic is sound, except, I don't want to return content in JSON. So how would I do it without doing that? Like a simple PHP `echo`?

**Comment by dereuromark** [comment]
Same thing, of course:
```php
    ...
    $this->autoRender = false;
    $data = 'OK';
    $this->response->body($data);
}

**Comment by pyrite357** [comment]
Hmm, seem to remember trying that and it not working. My site has existed since 1.1 by the way, so saying that I should have never done it in 2.x is kind of like, well, OK, but I didn't start out that way. I suppose much of my dilemma is my own fault (not having enough servers to run multiple cake versions on, not always using the latest and greatest methods of doing things, etc). I hope you guys don't think I am complaining, I love Cake, and just want to provide feedback.

**Comment by markstory** [comment]
I don't think we can reasonably make the response behave differently when autoRender is false because the response has no visibility into the controller. @pyrite357 If instead of an exception an error was logged would that work?

**Comment by dereuromark** [comment]
I wouldn't add such a snowflake behavior either. It shouldn't matter what kind of rendering it is.

**Comment by pyrite357** [comment]
@markstory I guess, I mean, we should probably ask the person who requested this "feature" to begin with. I think it was someone who said something about wanting to see what/if headers were sent. I can probably live with it just showing the Exception when debug is on, I'll just have to go turn debug off on all my methods. However, at one time, and this was long ago, I tried turning debug off in core, and some of my methods stopped working (I vaguely recall it being Prototype/ajax related). Ever since then, I just run my site at debug = 1, since it will only output extra stuff if an error occurs. However, in this case, it is always outputting an error, but doesn't say what the error is (Cake just says: An internal error has occurred). Well that doesn't help very much, haha. 

Anyway, I really appreciate you all for your consideration of the community regardless of what you decide to do. You all are the experts, and I'll work with whatever you come up with. If you all try to keep non-BC stuff out of minor releases, and not leave them undocumented like this one, that will help us out a lot!

**Comment by markstory** [comment]
Well this was intended to be backwards compatible. Given that is not the case reverting it to the point where is is backwards compatible is our best course of action.

Running CakePHP in production with debug on is not recommended as there are performance degradations and development errors enabled.

We never intended to put landmines like this in minor releases but mistakes happen. This will be corrected in 2.6.1 at least :-)

**Comment by markstory** [comment]
Closing as a pull request is up now.
